File: android/src/main/kotlin/app/tauri/serialplugin/manager/UsbPath.kt
package app.tauri.serialplugin.manager

/** Session path ↔ USB device name + port index (`deviceName#1` for multi-port). */
internal object UsbPath {
    private const val SEP = '#'

    fun sessionKey(deviceName: String, portIndex: Int, portCount: Int): String =
        if (portCount <= 1) deviceName else "$deviceName$SEP$portIndex"

    fun parse(path: String): Pair<String, Int> {
        val idx = path.lastIndexOf(SEP)
        if (idx < 0) return path to 0
        val device = path.substring(0, idx)
        val port = path.substring(idx + 1).toIntOrNull() ?: 0
        return device to port
    }

    fun deviceName(sessionPath: String): String = parse(sessionPath).first
}

File: src/android/usb_path.rs
//! Session path parsing (`deviceName` or `deviceName#port`).

pub fn parse(path: &str) -> (&str, usize) {
    if let Some(idx) = path.rfind('#') {
        let device = &path[..idx];
        let port = path[idx + 1..].parse().unwrap_or(0);
        (device, port)
    } else {
        (path, 0)
    }
}

pub fn session_key(device_name: &str, port_index: usize, port_count: usize) -> String {
    if port_count <= 1 {
        device_name.to_string()
    } else {
        format!("{device_name}#{port_index}")
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_single() {
        assert_eq!(parse("/dev/bus/usb/001/002"), ("/dev/bus/usb/001/002", 0));
    }

    #[test]
    fn parse_multi() {
        assert_eq!(parse("/dev/bus/usb/001/002#1"), ("/dev/bus/usb/001/002", 1));
    }
}
